tests/protocols/test_http.py

```diff
@@ -189,6 +189,12 @@ def get_connected_protocol(app, protocol_cls, event_loop, **kwargs):
         asyncio._set_running_loop(None)
 
 
+@pytest.fixture
+def event_loop():
+    with contextlib.closing(asyncio.new_event_loop()) as loop:
+        yield loop
```

Comment:
@euri10 this is introduced on this PR, jfyk

Reply:
the corresponding pytest-asyncio fixture `event_loop` is here

https://github.com/pytest-dev/pytest-asyncio/blob/4c7da65d6fcf9d725eccba28ad1ed2083524ee16/tests/async_fixtures/test_async_fixtures_with_finalizer.py#L19-L25

With this change, we don't close the loop after the yield, I'm surprised then pytest does not raise Resources not closed warnings, would be probably better to do it anyways so we dont get caught further down the road with such warning that are insanely hard to debug

and the scope is different, I'm not sure we need a loop for each test in that module, do we ?

Reply:
What is this fixture even used for here? It feels like an anti-pattern.

Reply:
> We do, the `contextlib.closing` calls `loop.close()`. Is that what you mean? thinking

yes, well if it passes tests then the ressources might be closed properly !

> What is this fixture even used for here? It feels like an anti-pattern.

iirc we use this pytest_asyncio fixture only in those "protocol" tests to create a bunch of mocks, mocked loop, mocked transport and then we create a protocol we interact with and look at its buffer,
is it an anti-pattern, idk :)

Reply:
Would be better to use a real event loop if that's what's used here

Reply:
I think this fixture should be removed. If it breaks protocol tests, we can surely update those tests.

Reply:
I'd be interested in seeing how, cant wrap my head around it, we need a loop to built the protocols, how can we get it if not mocked ?

Reply:
I might be misunderstanding something, we use pytest-asyncio event_loop fixture, which is to me a real event loop
```
@pytest.fixture
def event_loop(request):
    """Create an instance of the default event loop for each test case."""
    loop = asyncio.get_event_loop_policy().new_event_loop()
    yield loop
    loop.close()
```

we could avoid using the `MockLoop` doing something like this, is that what you mean ?

```diff
@@ -175,18 +175,15 @@ class MockTask:
 
 @contextlib.contextmanager
 def get_connected_protocol(app, protocol_cls, event_loop, **kwargs):
-    loop = MockLoop(event_loop)
-    asyncio._set_running_loop(loop)
     transport = MockTransport()
     config = Config(app=app, **kwargs)
     server_state = ServerState()
-    protocol = protocol_cls(config=config, server_state=server_state, _loop=loop)
+    protocol = protocol_cls(config=config, server_state=server_state, _loop=event_loop)
     protocol.connection_made(transport)
     try:
         yield protocol
     finally:
         protocol.loop.close()
-        asyncio._set_running_loop(None)
 
 
 @pytest.mark.parametrize("protocol_cls", HTTP_PROTOCOLS)
@@ -195,7 +192,8 @@ def test_get_request(protocol_cls, event_loop):
 
     with get_connected_protocol(app, protocol_cls, event_loop) as protocol:
         protocol.data_received(SIMPLE_GET_REQUEST)
-        protocol.loop.run_one()
+        tasks = asyncio.all_tasks(loop=event_loop)
+        protocol.loop.run_until_complete(*tasks)
         assert b"HTTP/1.1 200 OK" in protocol.transport.buffer
         assert b"Hello, world" in protocol.transport.buffer
```

Reply:
The loop is closed twice here (once in the fixture, once in `get_connected_protocol()`. Looks like the latter could be made into a fixture instead.

Does the protocol create tasks?

Reply:
as promised, here's the followup: https://github.com/encode/uvicorn/pull/1520